packages/xo-web/src/xo-app/vm/tab-advanced.js

```diff
@@ -573,6 +634,47 @@ export default class TabAdvanced extends Component {
     return createVusb(this.props.vm, pusb.usbGroup)
   }
 
+  _attachPcis = async () => {
+    const { vm } = this.props
+    const pcis = await confirm({
+      body: <PciAttachModal attachedPciIds={this._getAttachedPciIds()} pcisByHost={this.props.pcisByHost} vm={vm} />,
+      icon: 'add',
+      title: _('attachPcis'),
+    })
+    await vmAttachPcis(vm, pcis)
+  }
+
+  _getAttachedPciIds = createSelector(
+    () => this.props.vm,
+    vm => vm.other.pci?.split(',')?.map(s => s.split('/')[1]) ?? [] // other.pci: 0/pci_id,0/pci_id,..
+  )
+
+  _getPcis = createSelector(
+    () => this.props.vm,
+    () => this.props.pcisByHost,
+    (vm, pcisByHost) => {
+      if (!isVmRunning(vm)) {
+        // If the VM is not running, it's not attached to any host, therefore,
+        // we cannot determine which XAPI PCI object is associated with the given PCI_ID (eg: 0000:01:00.4).
+        // This determination depends on the specific host environment.
+        return {}
+      }
+      return keyBy(pcisByHost[vm.$container], 'pci_id')
+    }
+  )
+
+  _getIsPciPassthroughAvailable = () => {
+    const { pool, vm, vmHosts } = this.props
+
+    const host = vmHosts[pool?.master ?? vm.$container]
+    if (host === undefined) {
+      // Let's consider it `available` to avoid blocking a user who only has ACL on the VM
+      return true
```

Comment:
Maybe it would make sense to allow users to query this information about the host, then? Since you can get the information by just trying to attach a PCI to the VM anyway, if I understand correctly.